cinema/views.py

```
@@ -1,45 +1,109 @@
-from rest_framework.decorators import api_view
+from rest_framework import viewsets, mixins, generics, status
+from rest_framework.generics import get_object_or_404
 from rest_framework.response import Response
-from rest_framework import status
+from rest_framework.views import APIView
+from cinema.models import Movie, Actor, Genre, CinemaHall
+from cinema.serializers import (
+    MovieSerializer,
+    ActorSerializer,
+    GenreSerializer,
+    CinemaHallSerializer
+)
 
-from django.shortcuts import get_object_or_404
 
-from cinema.models import Movie
-from cinema.serializers import MovieSerializer
-
-
-@api_view(["GET", "POST"])
-def movie_list(request):
-    if request.method == "GET":
-        movies = Movie.objects.all()
-        serializer = MovieSerializer(movies, many=True)
+class GenreList(APIView):
+    def get(self, request) -> Response:
+        genres = Genre.objects.all()
+        serializer = GenreSerializer(genres, many=True)
         return Response(serializer.data, status=status.HTTP_200_OK)
 
-    if request.method == "POST":
-        serializer = MovieSerializer(data=request.data)
+    def post(self, request) -> Response:
+        serializer = GenreSerializer(data=request.data)
         if serializer.is_valid():
             serializer.save()
             return Response(serializer.data, status=status.HTTP_201_CREATED)
 
         return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```

You should raise an exception if the serializer is not valid as stated in the checklist item:

> 2. Don't forget return `Response` with `errors` if serializer is not valid:

Good example:

```python
class GenreList(APIView):
    def post(self, request):
        serializer = GenreSerializer(data=request.data)
        if serializer.is_valid():
            serializer.save()
            return Response(serializer.data, status=status.HTTP_201_CREATED)

        return Response(serializer.errors, status=status.HTTP_400_BAD_REQUEST)
```

Another good example:

```python
class GenreList(APIView):
    def post(self, request):
        serializer = GenreSerializer(data=request.data)
        serializer.is_valid(raise_exception=True)
        serializer.save()
        return Response(serializer.data, status=status.HTTP_201_CREATED)
```